sycl/doc/design/SYCLBINDesign.md

```diff
@@ -0,0 +1,289 @@
+# SYCLBIN - A format for separately compiled SYCL device code
+
+Some applications may want to dynamically load device binaries at runtime, e.g.
+for modularity and to avoid having to recompile the entire application. However,
+doing so through the
+[sycl_ext_oneapi_kernel_compiler](https://github.com/intel/llvm/blob/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc)
+extension can be unnecessarily expensive, if the systems utilizing this
+modularity are able to compile the binaries separate from the application's
+execution.
+
+To facilitate that a new SYCLBIN format is needed to define the interface between
+the compiler-produced binaries and the runtime's handling of it. This necessity
+comes from the overall design of our SYCL toolchain where runtime relies on
+compiler-provided information/metadata to implement various features (like
+support for specialization constants or shared libraries), i.e. device code alone
+is not enough.
+
+This design document details the SYCLBIN binary format used for storing SYCL
+device binaries to be loaded dynamically by the SYCL runtime. It also details
+how the toolchain produces, links and packages these binaries, as well as how
+the SYCL runtime library handles files of this format.
+
+
+## SYCLBIN binary format
+
+The files produced by the new compilation path will follow the format described
+in this section. The intention of defining a new format for these is to give
+the DPC++ implementation an extendable and lightweight wrapper around the
+multiple modules and corresponding metadata captured in the SYCLBIN file.
+The content of the SYCLBIN may be contained as an entry in the offloading binary
+format produced by the clang-offload-packager, as described in
+[ClangOffloadPackager.rst](https://github.com/intel/llvm/blob/sycl/clang/docs/ClangOffloadPackager.rst).
+
+The format of these files consist of a [file header](#file-header), containing
+general information about the file, followed by three lists of headers: The
+[abstract module header](#abstract-module-header) list, the
+[IR module header](#ir-module-header) list and
+[native device code image header](#native-device-code-image-header) list,
+containing information about the [abstract modules](#abstract-module),
+[IR modules](#ir-module) and
+[native device code images](#native-device-code-image) respectively.
+
+|                                                                       |
```

Comment:
This structure looks overcomplicated. Can be abstract modules, IR modules and device code images be unified in one entity like `offload_entry`? The type of the entity (SPIR-V, AOT and i.e) could be specified in a one global `image_table`. There already is a support of that in LLVM:
https://github.com/llvm/llvm-project/blob/c309abd92553191c404f2dc13f637bcfd53033f9/llvm/include/llvm/Frontend/Offloading/Utility.h#L27

My point here, is that we should strive to utilize the LLVM Offloading approach as much as we can and it was designed to be general.

Reply:
I don't understand how the linked `EntryTy` is relevant to this case? It talks about contents of a device image, but here we are talking about device images inside our own format. We could maybe combine the two entries, but it would make searching for e.g. native device code images over IR modules more difficult when going through the file and complexity-wise I don't see how it would be any different. We are very likely going to need different information for IR modules than for device binary images, so some distinction is necessary.